docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+let add x y = x + y
+let _ = map [||] [||] add
+```
+We will get an error:
+
+```bash
+let _ = map [||] [||] add
+                      ^^^
+This expression has type int -> int -> int
+but an expression was expected of type ('a -> 'b -> 'c) Js.Fn.arity2
+```
+
+To solve this, we add `@bs` in the function definition as well:
+
+```ocaml
+let add = fun [@bs] x y -> x + y
+```
+
+Annotating function definitions can be quite cumbersome when writing a lot of
+externals.
+
+To work around the verbosity, Melange offers another attribute called
+`bs.uncurry`.
+
+Let’s see how we could use it in the previous example. We just need to replace
+`bs` with `bs.uncurry`:
+
+```ocaml
+external map :
+  'a array -> 'b array -> (('a -> 'b -> 'c)[@bs.uncurry]) -> 'c array = "map"
+  [@@bs.val]
+```
+
+Now if we try to call `map` with a regular `add` function:
+
+```ocaml
+let add x y = x + y
+let _ = map [||] [||] add
+```
+
+Everything works fine now, without having to attach any attributes to `add`.
+
+The main difference between `bs` and `bs.uncurry` is that the latter only works
+with externals. `bs.uncurry` is the recommended option to use for bindings,
+while `bs` remains useful for those use cases where performance is crucial and
+we want the JavaScript functions generated from OCaml ones to not be applied
+partially.
+
+#### Modeling `this`-based Callbacks
+
+Many JavaScript libraries have callbacks which rely on the [`this`
+keyword](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this),
+for example:
+
+```js
+x.onload = function(v) {
+  console.log(this.response + v)
+}
+```
+
+Inside the `x.onload` callback, `this` would be pointing to `x`. It would not be
+correct to declare `x.onload` of type `unit -> unit`. Instead, Melange
+introduces a special attribute, `bs.this`, which allows to type `x` as this:
+
+```ocaml
+type x
+external x : x = "x" [@@bs.val]
+external set_onload : x -> ((x -> int -> unit)[@bs.this]) -> unit = "onload"
+  [@@bs.set]
+external resp : x -> int = "response" [@@bs.get]
+let _ =
+  set_onload x
+    begin
+      fun [@bs.this] o v -> Js.log (resp o + v)
+    end
+```
+
+Which generates:
+
+```javascript
+x.onload = function (v) {
+  var o = this;
+  console.log((o.response + v) | 0);
+};
+```
+
+Note that the first argument will be reserved for `this`.
+
+#### Wrapping returned nullable values
+
+For JavaScript functions that return a value that can be `undefined` or `null`,
+Melange provides `bs.return`. Using this attribute will have Melange generated
+code automatically convert the value returned by the function to an `option`
+type that can be used safely from Melange side, avoiding the need to use manual
+conversion functions like `Js.Nullable.toOption` and such.
+
+```ocaml
+type element
+type document
+external get_by_id : document -> string -> element option = "getElementById"
+  [@@bs.send] [@@bs.return nullable]
+
+let test document =
+  let elem = document |. get_by_id "header" in
```

Comment:
have we introduced `|.` before? This will be unfamiliar to people if we haven't.

Reply:
There is a "Chaining" section before this one that talks about it, but I agree better to remove it to avoid distractions. Done in https://github.com/melange-re/documentation-site/pull/12/commits/ce220da061bc231ee5c1cb4f6ab96a2ce0167746.